### PR TITLE
JAMES-3172 Fix task cancellation when started by Reactor

### PR DESCRIPTION
**Pull request by chibenwa**
Concensual part of https://github.com/linagora/james-project/pull/3356

### CODE REVIEW

**Comment by mbaechler** [comment]
Please take care of comments from https://github.com/linagora/james-project/pull/3356 here

**Comment by chibenwa** [comment]
> Please take care of comments from #3356 here

I think I did it.

**Comment by chibenwa** [comment]
That don't work. DistributedTaskManager tests fails.

**Comment by chibenwa** [comment]
Made it work with hard sweat.

I relaxed the tests a bit.

**Comment by chibenwa** [comment]
`[INFO] Apache James :: Server :: Mailbox :: JMX ........... FAILURE [  02:00 h]`

Sounds unrelated

**Comment by chibenwa** [comment]
No, it's related, and I **hate** mocks ;-)

File: server/task/task-memory/src/main/java/org/apache/james/task/SerialTaskManagerWorker.java [hunk 1]
```diff
@@ -107,30 +108,47 @@ public SerialTaskManagerWorker(Listener listener, Duration pollingInterval) {
...
-            .onFailure(() -> {
-                LOGGER.error("Task was partially performed. Check logs for more details. Taskid : " + taskWithId.getId());
-                Mono.from(listener.failed(taskWithId.getId(), taskWithId.getTask().details())).block();
-            });
+    private boolean isCausedByInterruptedException(Throwable e) {
+        if (e instanceof InterruptedException) {
+            return true;
+        }
+        return Optional.ofNullable(e.getCause())
+            .map(this::isCausedByInterruptedException)
```
Comment (mbaechler) [inline server/task/task-memory/src/main/java/org/apache/james/task/SerialTaskManagerWorker.java:124]: oh, a recursion, @blackheaven won't like it (:

What about 
```
        Stream.iterate(e, e.getCause() != null, e.getCause()).filter(e -> e instanceof InterruptedException).findAny();
```

File: server/task/task-memory/src/test/java/org/apache/james/task/SerialTaskManagerWorkerTest.java [hunk 2]
```diff
@@ -169,6 +168,39 @@ void theWorkerShouldReportThatATaskIsInProgress() throws InterruptedException {
         latch.countDown();
     }
 
+    @Test
+    void taskCallingReactorShouldDisposeReactorResourcesUponCancel() throws InterruptedException {
```
Comment (mbaechler) [inline server/task/task-memory/src/test/java/org/apache/james/task/SerialTaskManagerWorkerTest.java:172]:
```suggestion
    void taskExecutingReactivelyShouldStopExecutionUponCancel() throws InterruptedException {
```

File: server/task/task-memory/src/main/java/org/apache/james/task/SerialTaskManagerWorker.java (hunk 1 above)
Comment (chibenwa) [reply server/task/task-memory/src/main/java/org/apache/james/task/SerialTaskManagerWorker.java:124]: Arf, the call stack is expected to be small, I don't believe it makes much difference.

**Comment by chibenwa** [comment]
```
[181d0aa32b139ec0d6fdaa0463259a0fec27b885] Running: Shell Script
[181d0aa32b139ec0d6fdaa0463259a0fec27b885] [780] Running shell script
[181d0aa32b139ec0d6fdaa0463259a0fec27b885] + docker rm --force=true --volumes=true cassandra-cannot-cancel-light-ec2d78f2-9f62-458d-b1d3-df0d694c1c21 elasticSearch-cannot-cancel-light-ec2d78f2-9f62-458d-b1d3-df0d694c1c21 rabbitmq-cannot-cancel-light-ec2d78f2-9f62-458d-b1d3-df0d694c1c21 swift-cannot-cancel-light-ec2d78f2-9f62-458d-b1d3-df0d694c1c21 integration-cassandra-rabbitmq-cannot-cancel-light-ec2d78f2-9f62-458d-b1d3-df0d694c1c21 integration-cassandra-cannot-cancel-light-ec2d78f2-9f62-458d-b1d3-df0d694c1c21 integration-jpa-cannot-cancel-light-ec2d78f2-9f62-458d-b1d3-df0d694c1c21 integration-spring-cannot-cancel-light-ec2d78f2-9f62-458d-b1d3-df0d694c1c21 integration-memory-cannot-cancel-light-ec2d78f2-9f62-458d-b1d3-df0d694c1c21 james-parent-compile-cannot-cancel-light-ec2d78f2-9f62-458d-b1d3-df0d694c1c21 james-parent-test-cannot-cancel-light-ec2d78f2-9f62-458d-b1d3-df0d694c1c21 james-server-cassandra-cannot-cancel-light-ec2d78f2-9f62-458d-b1d3-df0d694c1c21 james-server-cassandra-rabbitmq-cannot-cancel-light-ec2d78f2-9f62-458d-b1d3-df0d694c1c21 james-server-cassandra-ldap-cannot-cancel-light-ec2d78f2-9f62-458d-b1d3-df0d694c1c21 james-server-jpa-cannot-cancel-light-ec2d78f2-9f62-458d-b1d3-df0d694c1c21 james-server-jpa-sample-testingcannot-cancel-light-ec2d78f2-9f62-458d-b1d3-df0d694c1c21 james-server-spring-cannot-cancel-light-ec2d78f2-9f62-458d-b1d3-df0d694c1c21 james-server-memory-cannot-cancel-light-ec2d78f2-9f62-458d-b1d3-df0d694c1c21 git-publish-cannot-cancel-light-ec2d78f2-9f62-458d-b1d3-df0d694c1c21
```

Still an unexpected failure...

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
```
[181d0aa32b139ec0d6fdaa0463259a0fec27b885] SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
[181d0aa32b139ec0d6fdaa0463259a0fec27b885] SLF4J: Defaulting to no-operation (NOP) logger implementation
[181d0aa32b139ec0d6fdaa0463259a0fec27b885] SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
[181d0aa32b139ec0d6fdaa0463259a0fec27b885] Error connecting to remote JMX agent : Failed to retrieve RMIServer stub: javax.naming.ServiceUnavailableException [Root exception is java.rmi.ConnectException: Connection refused to host: 127.0.0.1; nested exception is: 
[181d0aa32b139ec0d6fdaa0463259a0fec27b885] 	java.net.ConnectException: Connection refused (Connection refused)]
[181d0aa32b139ec0d6fdaa0463259a0fec27b885] usage: java org.apache.james.cli.ServerCmd --host <arg> <command>
[181d0aa32b139ec0d6fdaa0463259a0fec27b885]  -h,--host <arg>   node hostname or ip address
[181d0aa32b139ec0d6fdaa0463259a0fec27b885]  -p,--port <arg>   remote jmx agent port number
```

test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**